### PR TITLE
Support shrink to empty

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Follow up to https://github.com/apache/arrow-rs/pull/6790

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
FYI @emilk

File: arrow-buffer/src/buffer/mutable.rs [hunk 1]
```diff
@@ -468,7 +468,7 @@ impl MutableBuffer {
 }
 
 #[inline]
-fn dangling_ptr() -> NonNull<u8> {
+pub(crate) fn dangling_ptr() -> NonNull<u8> {
```
Comment (alamb) [inline arrow-buffer/src/buffer/mutable.rs:471]: I suggest adding some some comments about this function given its name.

Specifically, that it returns a valid, u8 pointer that is valid for the entire life of the process. Suitable for allocations of 0 length. 

I realize it wasn't introduced in this PR, but the name is confusing to me -- Maybe calling it `arbtrary_ptr()` or something would be less confusing 🤔

**Review by alamb — APPROVED** [review]
Thanks @tustvold  -- makes sense to me

File: arrow-buffer/src/buffer/mutable.rs (hunk 1 above)
Comment (tustvold) [reply arrow-buffer/src/buffer/mutable.rs:475]: Yeah dangling is an odd name, but is based on what the standard library calls it. I have added some docs to this effect

File: arrow-buffer/src/buffer/mutable.rs [hunk 2]
```diff
@@ -468,7 +468,7 @@ impl MutableBuffer {
...
     // SAFETY: ALIGNMENT is a non-zero usize which is then casted
     // to a *mut T. Therefore, `ptr` is not null and the conditions for
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow-buffer/src/buffer/mutable.rs]: I think this is a copy-pasta from `NonNull::dangling`

**Pull request merged**